### PR TITLE
adding DSL for jenkins jobs FLOC-4028

### PR DESCRIPTION
**Pull request by Azulinho**
Adds the DSL config for building the jenkins jobs for CI-slave-images

### CODE REVIEW

File: jobs.groovy [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+          regex: '(.*_(?i)aws_.*.*)' ],
+
+  on_rackspace:[description: 'All Rackspace Jobs',
+                regex: '(.*_(?i)rackspace_.*.*)' ],
+
+  on_centos_7:[description: 'All Centos Jobs',
+               regex: '(.*_(?i)Centos7)' ],
+
+  on_ubuntu_14_04_LTS:[description: 'All Ubuntu Trusty (14.04) Jobs',
+                       regex: '(.*_(?i)ubuntu1404)' ]
```
Comment (james-w) [inline jobs.groovy]: We will probably want a GCE view as well right?

File: jobs.groovy [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+// generate all the child jobs for our multijob
+for (cloud in on_clouds.keySet()) {
+  values = on_clouds.get(cloud)
+
+  for (region in values['regions']) {
+
+    for (distribution in values['distributions']) {
+
+      job_name = dashProject + '/' + dashBranchName + '/' + 'run_' +
+        cloud + '_' + region + '_' + distribution
```
Comment (james-w) [inline jobs.groovy]: There's a function defined earlier to help with this: full_job_name

File: jobs.groovy [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+
+          jobs_common_parameters.each { k, v ->
+            stringParam(k, v.default_value)
+          }
+
+          stringParam("CLOUD", cloud)
+          stringParam("DISTRIBUTION", distribution)
+          stringParam("AWS_REGION", region)
+          stringParam("OS_REGION_NAME", region)
+          stringParam("REGION", region)
```
Comment (james-w) [inline jobs.groovy]: Why are some of these not in jobs_common_parameters? They are used
in both places that map is.

File: jobs.groovy [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+            branch("${RECONFIGURE_BRANCH}")
+            clean(true)
+            createTag(false)
+            mergeOptions {
+              remote('upstream')
+              branch('master')
+              strategy('recursive')
+            }
+          }
+        }
```
Comment (james-w) [inline jobs.groovy]: We also retry the git clone in flocker.

File: jobs.groovy [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+        steps {
+          shell(with_steps)
+        }
+      }
+    }
+  }
+}
+
+// generate our multijob
+job_name = dashProject + '/' + dashBranchName + '/' + '__main_multijob'
```
Comment (james-w) [inline jobs.groovy]: You can use full_job_name here as well.

File: jobs.groovy [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+      for (cloud in on_clouds.keySet()) {
+        values = on_clouds.get(cloud)
+
+        for (region in values['regions']) {
+
+          for (distribution in values['distributions']) {
+
+            job_name = dashProject + '/' + dashBranchName + '/' + 'run_' +
+              cloud + '_' + region + '_' + distribution
```
Comment (james-w) [inline jobs.groovy]: full_job_name again.

**Comment by james-w** [comment]
Thanks for working on trying to get Jenkins to do this work for us. It will be a lot
nicer than having to run the fab script ourself and wait all day.

I have made a few comments on minor things that could be changed to improve
how maintainable this will be.

There could also be things like defining a function to yield the specific job configs,
rather than having the two places that do the nested loops to walk the structure.

Overall it looks ok, but I dislike that there are parts of this that are the same
as flocker, but we can't share the code between them. It means extra work
in applying improvements to both, and they aren't even in the same codebase.
It's only going to get worse as we add more projects as well. I don't have experience
of a better way to do what we are doing, but I know it's going to be painful
this way.

Merge when you have changed the points that you agree with.

Thanks,

James

File: jobs.groovy (hunk 1 above)
Comment (Azulinho) [inline jobs.groovy]: yeap, adding one

**Comment by Azulinho** [comment]
" dislike that there are parts of this that are the same
as flocker, but we can't share the code between them. It means extra work
in applying improvements to both, and they aren't even in the same codebase.
It's only going to get worse as we add more projects as well"

we can consume a set of common libraries both for 'bash' and for 'groovy' scripts that is deployed to the jenkins master and consumed by every Jenkins project.
But I'd rather leave that for another ticket

**Pull request merged**